cosi/cosi.md

```diff
@@ -258,64 +258,82 @@ steps below successfully.
-the participants N. The point A is defined as the collective key of the
-participants N. A collective signature is generated in four steps over two
+The leader must know the message M to be signed and the set of public keys of
+the participants P. The point A is defined as the collective key of the
+participants P. A collective signature is generated in four steps over two
 round trips between the leader and the rest of the participants. 
 
 ### Announcement
 
-Upon the request to generate a signature on a statement S, the leader broadcasts
-an announcement message indicating the start of a signing process. It is up to
-the implementation to decide whether to send S itself during that phase or not. 
+Upon the request to generate a signature on a message M, the leader broadcasts
+an Announcement packet indicating the start of a signing process. This
+Announcement packet MUST contain the message M to sign.
 
 ### Commitment
 
-Upon the receipt of an announcement message or if the participant is the leader,
-each participant i generates a random secret r_i by hashing 32 bytes of
+Upon the receipt of an Announcement packet, each non-leader participant SHOULD
+validate the message M syntactically and semantically according to an
+application-dependent policy. If any of these checks
+fails, the participant MUST abort the protocol.
+
+Each participant i then generates a random secret r_i by hashing 32 bytes of
 cryptographically secure random data. Each r_i MUST be re-generated until it is
 different from 0 mod L or 1 mod L. Each participants then constructs the
-commitment R_i as the encoding of [r_i]B, sends R_i to the leader and stores the
-generated r_i for usage in the response phase. If the participant is the leader,
-it executes the challenge step.
+commitment R_i as the encoding of [r_i]B, sends R_i in a Commitment packet to
+the leader and stores the generated r_i for usage in the response phase. If the
+participant is the leader, it executes the challenge step.
 
 ### Challenge
 
-The leader waits to receive the commitments R_i from the other participants for
-a certain time frame as defined by the application. After the timeout, the
-leader constructs the subset M of participants from whom he has received a
-commitment R_i and computes the sum R = SUM_{i in M}(R_i). The leader then
-computes SHA512(R || A || M) and interprets the resulting 64-byte digest as an
-integer c mod L.  The leader broadcasts c to all participants.
+The leader waits to receive the Commitment packets containing the R_i from the
+other participants for a certain time frame as defined by the application. After
+the timeout, the leader constructs the subset P' of participants from whom he
+has received a commitment R_i and computes the sum R = SUM_{i in P'}(R_i). The
+leader then computes SHA512(R || A || M) and interprets the resulting 64-byte
+digest as an integer c mod L.  The leader broadcasts c and R as a Challenge
```

Comment:
Why bother broadcasting c, when all cosigners can (and must anyway for security) recompute it anyway from R and the other inputs?